server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,345 @@
+    private final ReentrantReadWriteLock.WriteLock writeLock = readWriteLock.writeLock();
+    private final ReentrantReadWriteLock.ReadLock readLock = readWriteLock.readLock();
+
+    /**
+     * Constructor to initialise the composite directory
+     * @param localDirectory corresponding to the local FSDirectory
+     * @param remoteDirectory corresponding to the remote directory
+     * @param fileCache used to cache the remote files locally
+     */
+    public CompositeDirectory(FSDirectory localDirectory, RemoteSegmentStoreDirectory remoteDirectory, FileCache fileCache) {
+        super(localDirectory);
+        this.localDirectory = localDirectory;
+        this.remoteDirectory = remoteDirectory;
+        this.fileCache = fileCache;
+    }
+
+    /**
+     * Returns names of all files stored in this directory in sorted order
+     * Does not include locally stored block files (having _block_ in their names)
+     *
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public String[] listAll() throws IOException {
+        logger.trace("listAll() called ...");
+        readLock.lock();
+        try {
+            String[] localFiles = localDirectory.listAll();
+            logger.trace("LocalDirectory files : {}", () -> Arrays.toString(localFiles));
+            Set<String> allFiles = new HashSet<>(Arrays.asList(localFiles));
+            if (remoteDirectory != null) {
+                String[] remoteFiles = getRemoteFiles();
+                allFiles.addAll(Arrays.asList(remoteFiles));
+                logger.trace("Remote Directory files : {}", () -> Arrays.toString(remoteFiles));
+            }
+            Set<String> localLuceneFiles = allFiles.stream()
+                .filter(file -> !FileType.isBlockFile(file))
+                .collect(Collectors.toUnmodifiableSet());
+            String[] files = new String[localLuceneFiles.size()];
+            localLuceneFiles.toArray(files);
+            Arrays.sort(files);
+            logger.trace("listAll() returns : {}", () -> Arrays.toString(files));
+            return files;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * Removes an existing file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        logger.trace("deleteFile() called {}", name);
+        writeLock.lock();
+        try {
+            localDirectory.deleteFile(name);
+            fileCache.remove(localDirectory.getDirectory().resolve(name));
+        } catch (NoSuchFileException | FileNotFoundException e) {
+            logger.trace("File {} not found in local, trying to delete from Remote", name);
+            try {
+                remoteDirectory.deleteFile(name);
```

Comment:
Do we want to delete the files from `remoteDirectory` here? 

My understanding is that `remoteDirectory` is mutated only during commits.

Reply:
Updated in the latest revisions (we won't be deleting from the Remote Directory for now)